### ISSUE TITLE
Auth: Handle Exceptions in library

### ISSUE THREAD

**Issue by crtl** [kreait/firebase-php#370]
`Kreait\Firebase\Auth` class is documented to throw `Kreait\Firebase\Exception\FirebaseException`.
But actually the library does not handle internal exceptions.
For example when supplying invalid tokens and calling `verifyIdToken()` a `RuntimeException` or `LogicException` is thrown which should not be.

**Describe the solution you'd like**
Catch exceptions internally.

**Describe alternatives you've considered**
(Provided by @jeromegamez):
```
try {
    $firebaseAuth->verifyIdToken("123.123.123");
} catch (\Throwable $ex) {

}
```

Example:

```
try {
    $firebaseAuth->verifyIdToken("123.123.123");
} catch (FirebaseException $ex) {
  // Wont catch
}


{
"exception": "RuntimeException",
      "message": "Error while decoding to JSON: Malformed UTF-8 characters, possibly incorrectly encoded",
      "file": "D:\\dev\\htdocs\\project\\backend\\vendor\\lcobucci\\jwt\\src\\Parsing\\Decoder.php",
      "line": 36,
      "trace": [
        {
          "file": "D:\\dev\\htdocs\\project\\backend\\vendor\\lcobucci\\jwt\\src\\Parser.php",
          "line": 112,
          "function": "jsonDecode",
          "class": "Lcobucci\\JWT\\Parsing\\Decoder",
          "type": "->"
        },
        {
          "file": "D:\\dev\\htdocs\\project\\backend\\vendor\\lcobucci\\jwt\\src\\Parser.php",
          "line": 60,
          "function": "parseHeader",
          "class": "Lcobucci\\JWT\\Parser",
          "type": "->"
        },
        {
          "file": "D:\\dev\\htdocs\\project\\backend\\vendor\\kreait\\firebase-php\\src\\Firebase\\Auth\\IdTokenVerifier.php",
          "line": 130,
          "function": "parse",
          "class": "Lcobucci\\JWT\\Parser",
          "type": "->"
        },
        {
          "file": "D:\\dev\\htdocs\\project\\backend\\vendor\\kreait\\firebase-php\\src\\Firebase\\Auth\\IdTokenVerifier.php",
          "line": 50,
          "function": "ensureToken",
          "class": "Kreait\\Firebase\\Auth\\IdTokenVerifier",
          "type": "->"
        }
}
```

Sometimes an `Firebase\Auth\Token\Exception\UnknownKey` is thrown:

```json
{
"exception": "Firebase\\Auth\\Token\\Exception\\UnknownKey",
      "message": "A key with ID \"3494b1e786cdad092e423766bbe37f54ed87b22d\" could not be found.",
      "file": "D:\\dev\\htdocs\\project\\backend\\vendor\\kreait\\firebase-tokens\\src\\Firebase\\Auth\\Token\\Verifier.php",
      "line": 136,
      "trace": [
        {
          "file": "D:\\dev\\htdocs\\project\\backend\\vendor\\kreait\\firebase-tokens\\src\\Firebase\\Auth\\Token\\Verifier.php",
          "line": 68,
          "function": "getKey",
          "class": "Firebase\\Auth\\Token\\Verifier",
          "type": "->"
        },
        {
          "file": "D:\\dev\\htdocs\\project\\backend\\vendor\\kreait\\firebase-php\\src\\Firebase\\Auth\\IdTokenVerifier.php",
          "line": 53,
          "function": "verifyIdToken",
          "class": "Firebase\\Auth\\Token\\Verifier",
          "type": "->"
        },
        {
          "file": "D:\\dev\\htdocs\\project\\backend\\vendor\\kreait\\firebase-php\\src\\Firebase\\Auth.php",
          "line": 545,
          "function": "verifyIdToken",
          "class": "Kreait\\Firebase\\Auth\\IdTokenVerifier",
          "type": "->"
        },
        {
          "file": "D:\\dev\\htdocs\\project\\backend\\src\\Controller\\Api\\Account\\RegistrationController.php",
          "line": 118,
          "function": "verifyIdToken",
          "class": "Kreait\\Firebase\\Auth",
          "type": "->"
        },

```

**Comment by jeromegamez** [comment]
Thanks for bringing this up, you're right that not all possible exceptions are documented on the method in https://github.com/kreait/firebase-php/blob/master/src/Firebase/Auth.php#L521, and some are wrong - I will fix this.

However, internal exceptions should always be caught. For example, when I run

```php
try {
    $auth->verifyIdToken('123.123.123.123');
} catch (FirebaseException $e) {
    echo get_class($e).': '.$e->getMessage();
}
```

I get the following output

```
Kreait\Firebase\Exception\InvalidArgumentException: The given token could not be parsed: The JWT string must have two dots
```

Could you please show me the output of `composer show | grep kreait` and/or check that you the output contains this (or versions not too far off from this):

```
kreait/firebase-php                   4.38.0     Firebase Admin SDK
kreait/firebase-tokens                1.9.2      A library to work with Firebase tokens
```

**Comment by crtl** [comment]
```
$ composer show | grep kreait
kreait/clock                              1.0.1   A PHP 7.0 compatible clock abstraction
kreait/firebase-bundle                    1.8.0   Symfony Bundle for the Firebase Admin SDK
kreait/firebase-php                       4.38.0  Firebase Admin SDK
kreait/firebase-tokens                    1.9.1   A library to work with Firebase tokens
kreait/gcp-metadata                       1.0.1   Get the metadata from a Google Cloud Platform environment.

```

**Comment by jeromegamez** [comment]
Thanks for sharing! The versions look about right, so I can't explain why you're getting a RuntimeException instead of a more specific one.

For the other exceptions - I'm currently working on refinements on the token library, but I'm afraid you will not get around using more than one catch block - the ID token verification has a lot of legacies and I have to keep it as is to not break backward compatibility.

For the sake of simplicity (for now), I'd suggest using

```php
try {
    $firebaseAuth->verifyIdToken("123.123.123");
} catch (\Throwable $ex) {
  // Wont catch
}
```

If you want to handle specific cases, I hope that the better PHP Doc Blocks in the next release will help.

**Comment by crtl** [comment]
@jeromegamez In your example youve used the string `123.123.123.123` which is an invalid token format. Thats why you get an InvalidArgumentException. ``123.123.123` or any other JWT token in valid string format but as token invalid will throw an RuntimeException / LogicException.

**Comment by jeromegamez** [comment]
Aaah! Now, I get it too 😅 - thanks for clearing this up! I'll fix it!

**Comment by jeromegamez** [comment]
Could you please check, if the just-released [4.38.1](https://github.com/kreait/firebase-php/releases/tag/4.38.1) works for you?

All possible types of thrown exceptions are now documented (and only those should be thrown)

https://github.com/kreait/firebase-php/blob/c02500234e61abaa902724c7f4efbcdc2efcfb51/src/Firebase/Auth.php#L524-L529

and I updated the [related documentation](https://firebase-php.readthedocs.io/en/stable/authentication.html#verify-a-firebase-id-token) to include this usage example:

```php
use Firebase\Auth\Token\Exception\InvalidToken;
use Firebase\Auth\Token\Exception\UnknownKey;

// ...

try {
    $verifiedIdToken = $auth->verifyIdToken($idTokenString);
} catch (\InvalidArgumentException $e) {
    echo 'The token could not be parsed: '.$e->getMessage();
} catch (InvalidToken|UnknownKey $e) {
    echo 'The token is invalid: '.$e->getMessage();
}
```

I'll tackle the `InvalidToken`/`UnknownKey` ambiguation in a new release of the tokens library (which will then probably result in a new release of the SDK as well).

**Comment by crtl** [comment]
Thanks @jeromegamez ,
I will check it later this evening when im home.
